File: api/src/main/java/org/delivery/api/common/exception/ApiException.java
package org.delivery.api.common.exception;


import lombok.Getter;
import org.delivery.api.common.error.ErrorCodeIfs;

@Getter
public class ApiException extends RuntimeException implements ApiExceptionIfs{
    private final ErrorCodeIfs errorCodeIfs;
    private final String errorMessage;

    public ApiException(ErrorCodeIfs errorCodeIfs){
        super(errorCodeIfs.getMessage());
        this.errorCodeIfs = errorCodeIfs;
        this.errorMessage = errorCodeIfs.getMessage();
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, String errorDescription){
        super(errorDescription);
        this.errorCodeIfs = errorCodeIfs;
        this.errorMessage = errorDescription;
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, Throwable throwable){
        super(throwable);
        this.errorCodeIfs = errorCodeIfs;
        this.errorMessage = errorCodeIfs.getMessage();
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, Throwable throwable, String errorDescription){
        super(throwable);
        this.errorCodeIfs = errorCodeIfs;
        this.errorMessage = errorDescription;
    }
}

File: common/src/main/kotlin/org/delivery/common/exception/ApiException.kt
package org.delivery.common.exception

import org.delivery.common.error.ErrorCodeIfs

class ApiException : RuntimeException, ApiExceptionIfs{
    override val errorCodeIfs: ErrorCodeIfs
    override val errorMessage: String

    constructor(errorCodeIfs: ErrorCodeIfs): super(errorCodeIfs.getMessage()){
        this.errorCodeIfs = errorCodeIfs
        this.errorMessage = errorCodeIfs.getMessage()
    }
    constructor(errorCodeIfs: ErrorCodeIfs, errorDescription: String): super(errorDescription){
        this.errorCodeIfs = errorCodeIfs
        this.errorMessage = errorDescription
    }

    constructor(errorCodeIfs: ErrorCodeIfs, throwable: Throwable): super(throwable){
        this.errorCodeIfs = errorCodeIfs
        this.errorMessage = errorCodeIfs.getMessage()
    }

    constructor(errorCodeIfs: ErrorCodeIfs, throwable: Throwable, errorDescription: String): super(throwable){
        this.errorCodeIfs = errorCodeIfs
        this.errorMessage = errorDescription
    }

}